### ISSUE TITLE
reimplement in jax

### ISSUE THREAD

**Issue by nowittynamesleft** [nowittynamesleft/meta-nca#8]
pytorch is giving too many issues regarding in-place operations, with the dynamic computation graph making it difficult to debug/see precisely where the offending operations are (see #4 ). jax seems to make everything more explicit. should at least try to have a jax implementation to see these issues more clearly. also it might be faster (see #1 ).

**Comment by nowittynamesleft** [comment]
So far, local weight updates that apply across all weights every epoch without activations as input is implemented in jax version. Also implemented positional encoding instead of binary encoding. Currently able to get above 90% validation accuracy as of commit b8c8602 with 2 layer task network. 
To add:
- proportion of weights to not update per epoch
- activations of task network should be input to the local rule as perceptions of neighboring states
- train multiple architectures at once with local rule update

**Comment by nowittynamesleft** [comment]
also need to add wandb logging

**Comment by nowittynamesleft** [comment]
and weight visualization in addition to the hidden state visualization code

**Comment by nowittynamesleft** [comment]
Actually the good performance in commit b8c8602 is because of initializing in the same way every time, due to incorrect handling of random keys. Fixed, and performance is much worse for now.

Changed all instances of hidden_states to be (output_dim, input_dim, ...) like the weights originally are in jax. Removing transposes for weights, making all indexing/sum operations to handle this default (output_dim, input_dim) convention. The current commit c8435ed also adds animated GIFs of weights and their hidden states.

**Comment by nowittynamesleft** [comment]
Added proportion of weights to update per epoch (prop_cells_updated).

Major things still to do:
- split local rule updates by batches of inputs instead of complete dataset at once
- train multiple architectures at once with local rule update
- adding activations as part of perception vector

Minor improvements:
- Command line arguments
- add wandb logging
- split meta nca model code into its own file so that main run file can be simpler

**Comment by nowittynamesleft** [comment]
Multiple architecture training has been implemented in the jax code.

Other things besides the ones listed above:
- Time different parts of the code (compilation, task network update, metaepoch time)
- organize the code and make it easier to read/debug (especially concerning the loss and updating the task nets)
- optimize any for loops that are left if they are able to be replaced with jax.scan or jax.fori_loop calls

**Comment by nowittynamesleft** [comment]
more things needed:
- adding saving/loading model code!
- make sure it all works on gpu

**Comment by nowittynamesleft** [comment]
Implemented split task network updates by batches of inputs. Also removed extra commented out code to clean up a bit.
Training works on gpus, but is slower than CPU. Most likely sending info back and forth from GPU too many times. Need to optimize code for this.

**Comment by nowittynamesleft** [comment]
Added wandb logging previously. Reorganized a lot of code, put model code into regular_net_functions.py and metanca_functions.py files. Added saving/loading code.  

Still to be done:
- activations (however, I might leave this to later; can focus on running experiments on only having the perceptions be from neighboring weights and their hidden states)
- make command line arguments (and/or config file) for running main code. a config file may be preferable.
- optimizing the metanca model code. Most likely vectorizing the updates of all task net params with a single vmap function call on the local_rule_net_forward function, though need to think through if that's actually the best way to do it.
- Early stopping on the task_network_forward over epochs, for the test architecture evaluation

**Comment by nowittynamesleft** [comment]
also need to add:
- biases and their hidden states in the task networks
- separate evaluation script that takes in a local_rule_net_param hdf5 file that runs through different training scenarios/experiments

**Comment by nowittynamesleft** [comment]
added evaluation script and taking task net params from the best training loss among the epochs (pseudo-early stopping). added command line arguments to both evaluation script and jax_iris_nn.py and added randomly initialized task nets for comparisons.

**Comment by nowittynamesleft** [comment]
optimized the init_task_nets with vmap. This change seems to make training about 10x faster. But when timing the reset_task_nets portion and meta_nca_update portion of the metaepoch, the reset_task_nets portion still takes about 90% of the time of the metaepoch. I think there may be some other optimizations I can do to this function, maybe using scan to go through the for loop. Also, running on CPU is still maybe 10% faster than GPU.

**Comment by nowittynamesleft** [comment]
- evaluate script should plot best models
- checkpointing for local_rule_net to save lowest loss metamodel of metaepochs
- signals need to be between layers, so ALL connections between forward and back neurons, not just withinin the same layer. so that signals can travel between layers.

**Comment by nowittynamesleft** [comment]
signals are now between layers. adding the rest of the issues in this thread to the issues list and closing since all the previous functionality has been implemented + many improvements
